File: core/src/com/bombbird/terminalcontrol/screens/settingsscreen/categories/TrafficSettingsScreen.java
package com.bombbird.terminalcontrol.screens.settingsscreen.categories;

import com.badlogic.gdx.scenes.scene2d.Actor;
import com.badlogic.gdx.scenes.scene2d.ui.Image;
import com.badlogic.gdx.scenes.scene2d.ui.Label;
import com.badlogic.gdx.scenes.scene2d.ui.SelectBox;
import com.badlogic.gdx.scenes.scene2d.utils.ChangeListener;
import com.badlogic.gdx.utils.Align;
import com.badlogic.gdx.utils.Array;
import com.bombbird.terminalcontrol.TerminalControl;
import com.bombbird.terminalcontrol.entities.aircrafts.Emergency;
import com.bombbird.terminalcontrol.entities.airports.Airport;
import com.bombbird.terminalcontrol.screens.gamescreen.RadarScreen;
import com.bombbird.terminalcontrol.screens.settingsscreen.SettingsTab;
import com.bombbird.terminalcontrol.screens.settingsscreen.SettingsTemplateScreen;
import com.bombbird.terminalcontrol.utilities.saving.GameSaver;

import java.util.Locale;

public class TrafficSettingsScreen extends SettingsTemplateScreen {
    public SelectBox<String> emer;

    public Label emerChanceLabel;
    public Emergency.Chance emerChance;

    //In game only
    private SelectBox<String> tfcMode;
    private SelectBox<String> night;
    private SelectBox<String> nightStartHour;
    private SelectBox<String> nightStartMin;
    private SelectBox<String> nightEndHour;
    private SelectBox<String> nightEndMin;

    private Label tfcLabel;
    private RadarScreen.TfcMode tfcSel;

    private Label nightLabel;
    private Label timeLabel;
    private Label timeLabel2;

    private boolean allowNight;
    private int nightStart;
    private int nightEnd;

    public TrafficSettingsScreen(TerminalControl game, RadarScreen radarScreen, Image background) {
        super(game, radarScreen, background);

        loadUI(-1200, 0);

        setOptions();
    }

    /** Loads selectBox for display settings */
    @Override
    public void loadBoxes() {
        emer = createStandardSelectBox();
        emer.setItems("Off", "Low", "Medium", "High");
        emer.addListener(new ChangeListener() {
            @Override
            public void changed(ChangeEvent event, Actor actor) {
                emerChance = Emergency.Chance.valueOf(emer.getSelected().toUpperCase(Locale.US));
            }
        });

        if (radarScreen != null) {
            tfcMode = createStandardSelectBox();
            tfcMode.setItems("Normal", "Arrivals only");
            tfcMode.addListener(new ChangeListener() {
                @Override
                public void changed(ChangeEvent event, Actor actor) {
                    tfcSel = RadarScreen.TfcMode.valueOf(tfcMode.getSelected().toUpperCase(Locale.US).replaceAll(" ", "_"));
                }
            });

            night = createStandardSelectBox();
            night.setItems("On", "Off");
            night.addListener(new ChangeListener() {
                @Override
                public void changed(ChangeEvent event, Actor actor) {
                    allowNight = "On".equals(night.getSelected());
                    updateTabs(true);
                }
            });
            night.setName("night2");

            nightStartHour = new SelectBox<>(selectBoxStyle);
            Array<String> options7 = new Array<>(24);
            for (int i = 0; i < 24; i++) {
                String hour = Integer.toString(i);
                if (hour.length() == 1) hour = "0" + hour;
                options7.add(hour);
            }
            nightStartHour.setItems(options7);
            nightStartHour.setSize(300, 300);
            nightStartHour.setAlignment(Align.center);
            nightStartHour.getList().setAlignment(Align.center);
            nightStartHour.setName("night");

            nightStartMin = new SelectBox<>(selectBoxStyle);
            nightStartMin.setItems("00", "15", "30", "45");
            nightStartHour.addListener(new ChangeListener() {
                //Put here to prevent any potential NPE
                @Override
                public void changed(ChangeEvent event, Actor actor) {
                    nightStart = Integer.parseInt(nightStartHour.getSelected()) * 100 + Integer.parseInt(nightStartMin.getSelected());
                }
            });
            nightStartMin.addListener(new ChangeListener() {
                @Override
                public void changed(ChangeEvent event, Actor actor) {
                    nightStart = Integer.parseInt(nightStartHour.getSelected()) * 100 + Integer.parseInt(nightStartMin.getSelected());
                }
            });
            nightStartMin.setSize(300, 300);
            nightStartMin.setPosition(400, 0);
            nightStartMin.setAlignment(Align.center);
            nightStartMin.getList().setAlignment(Align.center);
            nightStartMin.setName("night");

            nightEndHour = new SelectBox<>(selectBoxStyle);
            nightEndHour.setItems(options7);
            nightEndHour.setSize(300, 300);
            nightEndHour.setPosition(950, 0);
            nightEndHour.setAlignment(Align.center);
            nightEndHour.getList().setAlignment(Align.center);
            nightEndHour.setName("night");

            nightEndMin = new SelectBox<>(selectBoxStyle);
            nightEndMin.setItems("00", "15", "30", "45");
            nightEndHour.addListener(new ChangeListener() {
                //Put here to prevent any potential NPE
                @Override
                public void changed(ChangeEvent event, Actor actor) {
                    nightEnd = Integer.parseInt(nightEndHour.getSelected()) * 100 + Integer.parseInt(nightEndMin.getSelected());
                }
            });
            nightEndMin.addListener(new ChangeListener() {
                @Override
                public void changed(ChangeEvent event, Actor actor) {
                    nightEnd = Integer.parseInt(nightEndHour.getSelected()) * 100 + Integer.parseInt(nightEndMin.getSelected());
                }
            });
            nightEndMin.setSize(300, 300);
            nightEndMin.setPosition(1350, 0);
            nightEndMin.setAlignment(Align.center);
            nightEndMin.getList().setAlignment(Align.center);
            nightEndMin.setName("night");
        }
    }

    /** Loads labels for display settings */
    @Override
    public void loadLabel() {
        super.loadLabel();
        emerChanceLabel = new Label("Emergencies: ", labelStyle);

        if (radarScreen != null) {
            tfcLabel = new Label("Traffic: ", labelStyle);

            nightLabel = new Label("Night mode: ", labelStyle);
            nightLabel.setName("night2");

            timeLabel = new Label("Active from:", labelStyle);
            timeLabel.setName("night");

            timeLabel2 = new Label("to", labelStyle);
            timeLabel2.setPosition(nightStartHour.getX() - nightLabel.getWidth() + 1300, nightStartHour.getY() + nightStartHour.getHeight() / 2 - timeLabel2.getHeight() / 2);
            timeLabel2.setName("night");
        }
    }

    /** Loads actors for display settings into tabs */
    @Override
    public void loadTabs() {
        SettingsTab tab1 = new SettingsTab(this, 2);
        tab1.addActors(emer, emerChanceLabel);

        if (radarScreen != null) {
            tab1.addActors(tfcMode, tfcLabel);
            tab1.addActors(night, nightLabel);
            tab1.addActors(nightStartHour, timeLabel, nightStartMin, nightEndHour, nightEndMin, timeLabel2);
        }

        settingsTabs.add(tab1);
    }

    /** Sets relevant options into select boxes */
    @Override
    public void setOptions() {
        if (radarScreen != null) {
            emerChance = radarScreen.emerChance;
        } else {
            emerChance = TerminalControl.emerChance;
        }
        String tmp = emerChance.toString().toLowerCase(Locale.US);
        emer.setSelected(tmp.substring(0, 1).toUpperCase() + tmp.substring(1));

        if (radarScreen != null) {
            tfcSel = radarScreen.tfcMode;
            String tmp2 = tfcSel.toString().toLowerCase(Locale.US);
            tfcMode.setSelected((tmp2.substring(0, 1).toUpperCase() + tmp2.substring(1)).replaceAll("_", " "));

            allowNight = radarScreen.allowNight;
            night.setSelected(allowNight ? "On" : "Off");

            nightStart = radarScreen.nightStart;
            String hr = Integer.toString(nightStart / 100);
            if (hr.length() == 1) hr = "0" + hr;
            nightStartHour.setSelected(hr);
            nightStartMin.setSelected(Integer.toString(nightStart % 100));

            nightEnd = radarScreen.nightEnd;
            String hr2 = Integer.toString(nightEnd / 100);
            if (hr2.length() == 1) hr2 = "0" + hr2;
            nightEndHour.setSelected(hr2);
            nightEndMin.setSelected(Integer.toString(nightEnd % 100));
        }
    }

    /** Confirms and applies the changes set */
    @Override
    public void sendChanges() {
        if (radarScreen == null) {
            TerminalControl.emerChance = emerChance;

            GameSaver.saveSettings();
        } else {
            radarScreen.emerChance = emerChance;
            radarScreen.tfcMode = tfcSel;

            boolean changed;
            changed = radarScreen.allowNight != allowNight || radarScreen.nightStart != nightStart || radarScreen.nightEnd != nightEnd;
            radarScreen.allowNight = allowNight;
            radarScreen.nightStart = nightStart;
            radarScreen.nightEnd = nightEnd;
            if (changed) {
                for (Airport airport: radarScreen.airports.values()) {
                    airport.updateRunwayUsage(); //Possible change in runway usage
                }
            }
            radarScreen.ui.updateInfoLabel();
        }
    }

    public boolean isAllowNight() {
        return allowNight && radarScreen != null;
    }
}

File: core/src/com/bombbird/terminalcontrol/screens/settingsscreen/categories/TrafficSettingsScreen.kt
package com.bombbird.terminalcontrol.screens.settingsscreen.categories

import com.badlogic.gdx.graphics.Color
import com.badlogic.gdx.scenes.scene2d.Actor
import com.badlogic.gdx.scenes.scene2d.ui.Image
import com.badlogic.gdx.scenes.scene2d.ui.Label
import com.badlogic.gdx.scenes.scene2d.ui.SelectBox
import com.badlogic.gdx.scenes.scene2d.ui.TextButton
import com.badlogic.gdx.scenes.scene2d.utils.ChangeListener
import com.badlogic.gdx.utils.Align
import com.badlogic.gdx.utils.Array
import com.bombbird.terminalcontrol.TerminalControl
import com.bombbird.terminalcontrol.entities.aircrafts.Emergency
import com.bombbird.terminalcontrol.screens.gamescreen.RadarScreen
import com.bombbird.terminalcontrol.screens.settingsscreen.SettingsTab
import com.bombbird.terminalcontrol.screens.settingsscreen.SettingsTemplateScreen
import com.bombbird.terminalcontrol.screens.settingsscreen.customsetting.TrafficFlowScreen
import com.bombbird.terminalcontrol.utilities.Fonts
import com.bombbird.terminalcontrol.utilities.saving.GameSaver
import java.util.*

class TrafficSettingsScreen(game: TerminalControl, radarScreen: RadarScreen?, background: Image?) : SettingsTemplateScreen(game, radarScreen, background) {
    lateinit var emer: SelectBox<String>
    lateinit var emerChanceLabel: Label
    lateinit var emerChance: Emergency.Chance

    //In game only
    private lateinit var tfcMode: SelectBox<String>
    private lateinit var night: SelectBox<String>
    private lateinit var nightStartHour: SelectBox<String>
    private lateinit var nightStartMin: SelectBox<String>
    private lateinit var nightEndHour: SelectBox<String>
    private lateinit var nightEndMin: SelectBox<String>
    private lateinit var flowButton: TextButton
    private lateinit var tfcLabel: Label
    private lateinit var tfcSel: RadarScreen.TfcMode
    private lateinit var nightLabel: Label
    private lateinit var timeLabel: Label
    private lateinit var timeLabel2: Label
    private var allowNight = false
    private var nightStart = 0
    private var nightEnd = 0

    init {
        infoString = "Set the traffic settings below."
        loadUI(-1200, -200)
        setOptions()
        updateTabs(true)
    }

    /** Loads selectBox for display settings  */
    override fun loadBoxes() {
        emer = createStandardSelectBox()
        emer.setItems("Off", "Low", "Medium", "High")
        emer.addListener(object : ChangeListener() {
            override fun changed(event: ChangeEvent, actor: Actor) {
                emerChance = Emergency.Chance.valueOf(emer.selected.toUpperCase(Locale.US))
            }
        })

        if (radarScreen != null) {
            tfcMode = createStandardSelectBox()
            tfcMode.setItems("Normal", "Arrivals only")
            tfcMode.addListener(object : ChangeListener() {
                override fun changed(event: ChangeEvent, actor: Actor) {
                    tfcSel = RadarScreen.TfcMode.valueOf(tfcMode.selected.toUpperCase(Locale.US).replace(" ".toRegex(), "_"))
                }
            })
            night = createStandardSelectBox()
            night.setItems("On", "Off")
            night.addListener(object : ChangeListener() {
                override fun changed(event: ChangeEvent, actor: Actor) {
                    allowNight = "On" == night.selected
                    updateTabs(true)
                }
            })
            night.name = "night2"
            nightStartHour = SelectBox(selectBoxStyle)
            val options7 = Array<String>(24)
            for (i in 0..23) {
                var hour = i.toString()
                if (hour.length == 1) hour = "0$hour"
                options7.add(hour)
            }
            nightStartHour.items = options7
            nightStartHour.setSize(300f, 300f)
            nightStartHour.setAlignment(Align.center)
            nightStartHour.list.setAlignment(Align.center)
            nightStartHour.name = "night"
            nightStartMin = SelectBox(selectBoxStyle)
            nightStartMin.setItems("00", "15", "30", "45")
            nightStartHour.addListener(object : ChangeListener() {
                //Put here to prevent any potential NPE
                override fun changed(event: ChangeEvent, actor: Actor) {
                    nightStart = nightStartHour.selected.toInt() * 100 + nightStartMin.selected.toInt()
                }
            })
            nightStartMin.addListener(object : ChangeListener() {
                override fun changed(event: ChangeEvent, actor: Actor) {
                    nightStart = nightStartHour.selected.toInt() * 100 + nightStartMin.selected.toInt()
                }
            })
            nightStartMin.setSize(300f, 300f)
            nightStartMin.setPosition(400f, 0f)
            nightStartMin.setAlignment(Align.center)
            nightStartMin.list.setAlignment(Align.center)
            nightStartMin.name = "night"
            nightEndHour = SelectBox(selectBoxStyle)
            nightEndHour.items = options7
            nightEndHour.setSize(300f, 300f)
            nightEndHour.setPosition(950f, 0f)
            nightEndHour.setAlignment(Align.center)
            nightEndHour.list.setAlignment(Align.center)
            nightEndHour.name = "night"
            nightEndMin = SelectBox(selectBoxStyle)
            nightEndMin.setItems("00", "15", "30", "45")
            nightEndHour.addListener(object : ChangeListener() {
                //Put here to prevent any potential NPE
                override fun changed(event: ChangeEvent, actor: Actor) {
                    nightEnd = nightEndHour.selected.toInt() * 100 + nightEndMin.selected.toInt()
                }
            })
            nightEndMin.addListener(object : ChangeListener() {
                override fun changed(event: ChangeEvent, actor: Actor) {
                    nightEnd = nightEndHour.selected.toInt() * 100 + nightEndMin.selected.toInt()
                }
            })
            nightEndMin.setSize(300f, 300f)
            nightEndMin.setPosition(1350f, 0f)
            nightEndMin.setAlignment(Align.center)
            nightEndMin.list.setAlignment(Align.center)
            nightEndMin.name = "night"
        }
    }

    /** Loads labels for display settings  */
    override fun loadLabel() {
        super.loadLabel()
        emerChanceLabel = Label("Emergencies: ", labelStyle)
        if (radarScreen != null) {
            tfcLabel = Label("Traffic: ", labelStyle)
            nightLabel = Label("Night mode: ", labelStyle)
            nightLabel.name = "night2"
            timeLabel = Label("Active from:", labelStyle)
            timeLabel.name = "night"
            timeLabel2 = Label("to", labelStyle)
            timeLabel2.setPosition(nightStartHour.x - nightLabel.width + 1300, nightStartHour.y + nightStartHour.height / 2 - timeLabel2.height / 2)
            timeLabel2.name = "night"
        }
    }

    override fun loadButton() {
        super.loadButton()
        val textButtonStyle = TextButton.TextButtonStyle()
        textButtonStyle.font = Fonts.defaultFont20
        textButtonStyle.fontColor = Color.WHITE
        textButtonStyle.up = TerminalControl.skin.getDrawable("Button_up")
        textButtonStyle.down = TerminalControl.skin.getDrawable("Button_down")
        flowButton = TextButton("Arrival traffic settings", textButtonStyle)
        flowButton.setSize(1200f, 300f)
        flowButton.addListener(object : ChangeListener() {
            override fun changed(event: ChangeEvent, actor: Actor) {
                //New traffic page
                game.screen = TrafficFlowScreen(game)
                event.handle()
            }
        })
    }

    /** Loads actors for display settings into tabs  */
    override fun loadTabs() {
        val tab1 = SettingsTab(this, 2)
        tab1.addActors(emer, emerChanceLabel)
        if (radarScreen != null) {
            tab1.addActors(tfcMode, tfcLabel)
            tab1.addActors(night, nightLabel)
            tab1.addActors(nightStartHour, timeLabel, nightStartMin, nightEndHour, nightEndMin, timeLabel2)
            tab1.addButton(flowButton)
        }
        settingsTabs.add(tab1)
    }

    /** Sets relevant options into select boxes  */
    override fun setOptions() {
        emerChance = radarScreen?.emerChance ?: TerminalControl.emerChance
        val tmp = emerChance.toString().toLowerCase(Locale.US)
        emer.selected = tmp.substring(0, 1).toUpperCase(Locale.ROOT) + tmp.substring(1)
        if (radarScreen != null) {
            tfcSel = radarScreen.tfcMode
            val tmp2 = tfcSel.toString().toLowerCase(Locale.US)
            tfcMode.selected = (tmp2.substring(0, 1).toUpperCase(Locale.ROOT) + tmp2.substring(1)).replace("_".toRegex(), " ")
            allowNight = radarScreen.allowNight
            night.selected = if (allowNight) "On" else "Off"
            nightStart = radarScreen.nightStart
            var hr = (nightStart / 100).toString()
            if (hr.length == 1) hr = "0$hr"
            nightStartHour.selected = hr
            nightStartMin.selected = (nightStart % 100).toString()
            nightEnd = radarScreen.nightEnd
            var hr2 = (nightEnd / 100).toString()
            if (hr2.length == 1) hr2 = "0$hr2"
            nightEndHour.selected = hr2
            nightEndMin.selected = (nightEnd % 100).toString()
        }
    }

    /** Confirms and applies the changes set  */
    override fun sendChanges() {
        if (radarScreen == null) {
            TerminalControl.emerChance = emerChance
            GameSaver.saveSettings()
        } else {
            radarScreen.emerChance = emerChance
            radarScreen.tfcMode = tfcSel
            val changed = radarScreen.allowNight != allowNight || radarScreen.nightStart != nightStart || radarScreen.nightEnd != nightEnd
            radarScreen.allowNight = allowNight
            radarScreen.nightStart = nightStart
            radarScreen.nightEnd = nightEnd
            if (changed) {
                for (airport in radarScreen.airports.values) {
                    airport.updateRunwayUsage() //Possible change in runway usage
                }
            }
            radarScreen.ui.updateInfoLabel()
        }
    }

    fun isAllowNight(): Boolean {
        return allowNight && radarScreen != null
    }
}